### PR TITLE
task: public api with pat

### PR DESCRIPTION
**Pull request by KubaBaniak**
### Acceptance criteria:
- User should be able to generate his personal access token (PAT)
- User should be able to use his PAT instead of Bearer token to access endpoints for:
   - fetching list or recipes
   - fetching a specific recipe
   - updating a recipe
   - deleting a recipe
- Generated PAT should not expire

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -36,7 +36,28 @@ export class AuthService {
...
+    return await this.jwtService.signAsync(
+      {
+        id: user.id,
+        email: user.email,
+      },
+      { expiresIn: `${process.env.JWT_EXPIRY_TIME}s` },
+    );
+  }
+
+  async createPAT(userId: number): Promise<string> {
```
Comment (rad20c) [inline src/auth/auth.service.ts:48]: I'd prefer using full name -> `Personal Access Token` here

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -36,7 +36,28 @@ export class AuthService {
...
+    const validPatToken = await this.userRepository.getUserValidPatWithId(
```
Comment (rad20c) [inline src/auth/auth.service.ts:49]:
```suggestion
    const validPatToken = await this.userRepository.getValidPatForUserId(
```

Comment (rad20c) [reply src/auth/auth.service.ts:49]: `PatToken` sounds strange -> "Personal Access Token Token"

File: src/auth/dto/create-pat-response.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export class CreatePatResponse {
+  public patToken: string;
+
+  constructor(patToken: string) {
```
Comment (rad20c) [inline src/auth/dto/create-pat-response.ts:4]: patToken again - I'd just name it personalAccessToken

File: prisma/migrations/20230811101255_add_pta/migration.sql [hunk 4]
```diff
@@ -0,0 +1,14 @@
+-- CreateTable
+CREATE TABLE "PTA" (
```
[the remark this answers is not in the document]
Comment (rad20c) [reply prisma/migrations/20230811101255_add_pta/migration.sql:2]: I would name the table `PersonalAccessTokens`

File: example.env [hunk 5]
```diff
@@ -5,6 +5,7 @@ PGADMIN_EMAIL=
 PGADMIN_PASSWORD=
 JWT_EXPIRY_TIME=
 JWT_SECRET=
+JWT_EXPIRY_TIME=
```
Comment (rad20c) [inline example.env:8]: duplicated entry `JWT_EXPIRY_TIME`

File: docs/CHANGELOG.md
Comment (rad20c) [inline docs/CHANGELOG.md:1]: hmm, this file should not be in this PR (there should be no changes to this file in the PR)

File: prisma/migrations/20230811101958_rename_to_pat/migration.sql [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+*/
+-- DropForeignKey
+ALTER TABLE "PTA" DROP CONSTRAINT "PTA_userId_fkey";
+
+-- DropTable
+DROP TABLE "PTA";
+
+-- CreateTable
+CREATE TABLE "PAT" (
```
Comment (rad20c) [inline prisma/migrations/20230811101958_rename_to_pat/migration.sql:14]: ok, now I see - but again I would name it `PersonalAccessTokens` - when you dive in into the code for the first time you might not know if PAT is PersonalAccessToken or Port Address Translation or Portland Association of Teachers

File: prisma/schema.prisma [hunk 7]
```diff
@@ -28,7 +29,15 @@ model Recipe {
...
-  imageKeys  String[] @unique
+  imageKeys   String[] @unique
+}
+
+model PAT {
+  id            Int       @id @default(autoincrement())
+  user          User      @relation(fields: [userId], references: [id])
+  userId        Int
+  token         String
+  invalidatedAt DateTime?
```
Comment (rad20c) [inline prisma/schema.prisma:40]: I'd also add `createdAt` field

File: src/auth/auth.module.ts [hunk 8]
```diff
@@ -8,6 +8,7 @@ import { LocalStrategy } from './strategies/local.strategy';
 import { JwtStrategy } from './strategies/jwt.strategy';
 import { PassportModule } from '@nestjs/passport';
 import { LocalAuthGuard } from './guards/local-auth.guard';
+import { PtaJwtStrategy } from './strategies/pta-jwt.strategy';
```
Comment (rad20c) [inline src/auth/auth.module.ts:11]: pta again

File: src/auth/auth.service.ts [hunk 9]
```diff
@@ -36,7 +36,28 @@ export class AuthService {
       throw new NotFoundException();
     }
 
-    return await this.jwtService.signAsync({ id: user.id, email: user.email });
+    return await this.jwtService.signAsync(
```
Comment (rad20c) [inline src/auth/auth.service.ts:39]: no need to await on return

File: src/auth/auth.service.ts [hunk 10]
```diff
@@ -36,7 +36,28 @@ export class AuthService {
...
+    );
+    if (validPatToken) {
+      this.userRepository.invalidateUserPat(userId);
+    }
+    const token = await this.jwtService.signAsync({
+      id: userId,
+      type: 'PAT',
+    });
+    const patToken = await this.userRepository.createPat(userId, token);
+    return patToken.token;
```
Comment (rad20c) [inline src/auth/auth.service.ts:60]:
```suggestion
    const { token } = await this.userRepository.createPat(userId, token);
    return token;
```

I think it's more clear. `patToken.token` sounded strange

Comment (rad20c) [reply src/auth/auth.service.ts:60]: I'd rename `createPat` to `createPersonalAccessToken` or `generatePersonalAccessToken`

File: src/auth/guards/pta-auth-api.guard.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+import { Injectable } from '@nestjs/common';
+import { AuthGuard } from '@nestjs/passport';
+
+@Injectable()
+export class ApiAuthGuard extends AuthGuard(['jwt.pta', 'jwt.bearer']) {}
```
Comment (rad20c) [inline src/auth/guards/pta-auth-api.guard.ts:5]: pta?

File: src/auth/strategies/pta-jwt.strategy.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import { ExtractJwt, Strategy } from 'passport-jwt';
+import { PassportStrategy } from '@nestjs/passport';
+import { Injectable } from '@nestjs/common';
+import { jwtConstants } from '../constants';
+import { PtaJwtPayload } from './payloads/pta-jwt-token.payload';
+
+@Injectable()
+export class PtaJwtStrategy extends PassportStrategy(Strategy, 'jwt.pta') {
```
Comment (rad20c) [inline src/auth/strategies/pta-jwt.strategy.ts:8]: I'd name it `PersonalAccessTokenStrategy`

File: src/auth/strategies/jwt.strategy.ts [hunk 13]
```diff
@@ -5,11 +5,10 @@ import { jwtConstants } from '../constants';
 import { JwtPayload } from './payloads/jwt-token.payload';
 
 @Injectable()
-export class JwtStrategy extends PassportStrategy(Strategy) {
+export class JwtStrategy extends PassportStrategy(Strategy, 'jwt.bearer') {
```
Comment (rad20c) [inline src/auth/strategies/jwt.strategy.ts:8]: I think we can rename it to e.g. `BearerStrategy`

File: src/auth/strategies/local.strategy.ts [hunk 14]
```diff
@@ -15,7 +15,7 @@ export class LocalStrategy extends PassportStrategy(Strategy) {
     const user = await this.authService.validateUser(data);
 
     if (!user) {
-      throw new UnauthorizedException();
+      throw new NotFoundException();
```
Comment (rad20c) [inline src/auth/strategies/local.strategy.ts:18]: I think the `UnauthorizedException()` is an appropriate error here. Now we are explicitly telling user that "hey, this user does not exist in the system", thus enabling malicious actors to perform a user enumeration attack. Note that `this.authService.validateUser(data);` method can return `UnauthorizedException()` on its own. We should return `UnauthorizedException()` here

File: src/auth/strategies/pta-jwt.strategy.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  constructor() {
+    super({
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+      secretOrKey: jwtConstants.secret,
+    });
+  }
+  validate(payload: PtaJwtPayload): PtaJwtPayload {
```
Comment (rad20c) [inline src/auth/strategies/pta-jwt.strategy.ts:15]: pta?

File: src/metadata.ts [hunk 16]
```diff
@@ -153,6 +164,9 @@ export default async () => {
             AuthController: {
               signIn: { type: t['./auth/dto/sign-in-response'].SignInResponse },
               signUp: { type: t['./auth/dto/sign-up-response'].SignUpResponse },
+              generatePAT: {
+                type: t['./auth/dto/create-pat-response'].CreatePatResponse,
```
Comment (rad20c) [inline src/metadata.ts:168]: hmm, where did `generatePAT` came from? You have `createPat` / `createPAT` everywhere

File: src/user/user.repository.ts [hunk 17]
```diff
@@ -48,4 +47,40 @@ export class UserRepository {
...
+      data: {
+        token,
+        user: {
+          connect: {
+            id: userId,
+          },
+        },
+      },
+    });
+    return test;
```
Comment (rad20c) [inline src/user/user.repository.ts:62]: you should be able to immediately return here. Also I think `test` field was for test purposes right? 😉

File: src/user/user.repository.ts [hunk 18]
```diff
@@ -48,4 +47,40 @@ export class UserRepository {
...
+        userId,
+        invalidatedAt: {
+          equals: null,
+        },
+      },
+    });
+  }
+
+  async invalidateUserPat(userId: number): Promise<void> {
+    await this.prisma.pAT.updateMany({
```
Comment (rad20c) [inline src/user/user.repository.ts:77]: you can remove `async` and `await` and keep `Promise<void>` - you will still be able to `await` it where you call this method

File: test/recipe.e2e-spec.ts [hunk 19]
```diff
@@ -24,6 +24,7 @@ describe('RecipeController (e2e)', () => {
   let authService: AuthService;
   let user: User;
   let accessToken: string;
+  let patToken: string;
```
Comment (rad20c) [inline test/recipe.e2e-spec.ts:27]: "patToken" sound strange - I'd rename to `personalAccessToken`

**Review by rad20c — CHANGES_REQUESTED** [review]
check comments

File: src/metadata.ts (hunk 16 above)
Comment (KubaBaniak) [reply src/metadata.ts:168]: this is the name of controller method

File: src/user/user.repository.ts (hunk 18 above)
Comment (KubaBaniak) [reply src/user/user.repository.ts:77]: I've tested it out and without `async` `await` in `user.repository` it doesn't update `invalidatedAt` field (I awaited this in `auth.service`).

File: src/auth/strategies/local.strategy.ts (hunk 14 above)
Comment (KubaBaniak) [reply src/auth/strategies/local.strategy.ts]: Ok I get it. If so then should I also change the Exception within `signIn` method in `auth.service`?

File: prisma/schema.prisma [hunk 20]
```diff
@@ -11,11 +11,12 @@ datasource db {
...
-  email    String   @unique
-  password String
-  role     Role     @default(USER)
-  recipes  Recipe[]
+  id                  Int                   @id @default(autoincrement())
+  email               String                @unique
+  password            String
+  role                Role                  @default(USER)
+  recipes             Recipe[]
+  personalAccessToken PersonalAccessToken[]
```
Comment (rad20c) [inline prisma/schema.prisma:19]: hmm, I wouldn't add `personalAccessToken` to the `User` table. You already have `userId` in `PersonalAccessToken` table - this should be enough.

File: src/auth/strategies/local.strategy.ts (hunk 14 above)
Comment (rad20c) [reply src/auth/strategies/local.strategy.ts]: if the logic is exactly the same as above, then yes, I would change it 👍🏻

File: src/user/user.repository.ts (hunk 18 above)
Comment (rad20c) [reply src/user/user.repository.ts]: hmm, have you tried returning it? I mean adding `return` statement.

However, we might not want to return anything (so leaving `Promise<void>` intact) -in this case what you have now is correct 👍🏻

File: src/metadata.ts (hunk 16 above)
Comment (rad20c) [reply src/metadata.ts:168]: ah I see - how about we unify the method name for creating the personal access token everywhere (in controller and service)?

File: src/auth/auth.service.ts (hunk 10 above)
Comment (rad20c) [reply src/auth/auth.service.ts]: Regarding suggestion - sure, but I'd remove `ToDb` part.

But now I'm thinking - why do we have `createPat` method on `userRepository`? Shouldn't we call `personalAccessTokenRepository` instead?

File: prisma/schema.prisma (hunk 20 above)
Comment (KubaBaniak) [reply prisma/schema.prisma:19]: I can't remove `personalAccessToken` here.
Prisma throws this error:
```
Error validating field 'user' in model 'PersonalAccessToken': 
The relation field 'user' on model 'PersonalAccessToken' is missing an opposite relation field on the model 'User'`
```

**Review by rad20c — APPROVED** [review]

**Pull request merged**